Better NUMA behavior

wingo opened the pull request:
This branch makes it so that the --cpu argument also binds memory to a NUMA node.  It also checks that the PCI cards being used are on the same NUMA node as the CPU, as well as troubleshooting some other NUMA-related issues.

It mostly makes it unnecessary to use taskset/numactl.  The exception is that pages that are already allocated when the program starts aren't migrated to the new NUMA node, if they were allocated on the "wrong" NUMA node.  Not sure what to do about this yet.

wingo committed:
Add get_mempolicy, set_mempolicy NUMA syscalls

Calling get_mempolicy() will return an object with "mode" and "mask"
keys, corresponding to the two output arguments of the get_mempolicy
function. The mask is implemented along the lines of cpu_set, except
that it can hold any number of bits, defaulting to the size of a long.

set_mempolicy(mode, mask) imposes a mode and possibly a mask as well.

wingo committed:
Add beginnings of NUMA module for Snabb

wingo committed:
Add cpu_get_numa_node

wingo committed:
Add more NUMA methods

wingo committed:
Add choose_numa_node_for_pci_addresses

wingo committed:
pci_get_numa_node can accept canonical addresses

wingo committed:
Add numa.check_affinity_for_pci_addresses

wingo committed:
Use lib.numa in lwaftr run

wingo committed:
Work around a broken `getcpu'

kbara commented:
Awesome work; I'm glad to see this being simplified (from a user perspective) - it should help us run more consistent tests, and cut down on deployment complexity. How confident are you that the level of testing contained in this patch is robust enough to make sure it continues to work?

@dpino any idea why snabb1 hasn't run the test?

kbara commented:
https://github.com/justincormack/ljsyscall/pull/195 fixes getcpu, btw.

dpino commented:
LGTM.

wingo committed:
ljsyscall: Fix getcpu()

Patch by Katerina Barone-Adesi <kbarone@igalia.com>, committed upstream
via https://github.com/justincormack/ljsyscall/pull/195.

wingo committed:
ljsyscall: Add migrate_pages.

* lib/ljsyscall/syscall/linux/c.lua:
* lib/ljsyscall/syscall/linux/syscalls.lua: Add support for the
  migrate_pages Linux syscall.

wingo committed:
numa: Use ljsyscall's fixed "getcpu()" implementation

* src/lib/numa.lua: Rely on ljsyscall's getcpu(), now that it's fixed.

wingo committed:
numa: Migrate existing pages when binding to NUMA node.

* src/lib/numa.lua:

wingo committed:
lwaftr loadtest uses lib.numa

* src/program/lwaftr/loadtest/loadtest.lua: Add --cpu and NUMA
  diagnostics.

kbara commented:
LGTM

wingo commented:
I cherry-picked in the fix for getcpu and added a wrapper for migrate_pages, to fix the problem that there might be pages that already have the wrong affinity.  It appears to work!
